### ISSUE TITLE
False positive: "mismatched types [E0308]" on macro-implemented traits

### ISSUE THREAD

**Issue by noneuclideangirl** [intellij-rust/intellij-rust#5240]
Hi all, this is my first time submitting a public issue so let me know if I can make it more helpful. I've noticed a particularly annoying spurious error related to traits implemented by macros. I'm not entirely sure whether the Rust linter is supposed to evaluate macros from crates and use that evaluation to inform its checking, but I do think it should at some point if it doesn't already.

## Environment

* **IntelliJ Rust plugin version:** `0.2.120.2175-193`
* **Rust toolchain version:**
```
rustup 1.21.1 (7832b2ebe 2019-12-20)
rustc 1.44.0-nightly (e82734e56 2020-04-11)
cargo 1.44.0-nightly (390e8f245 2020-04-07)
```
* **IDE name and version:** CLion 2019.3.5
* **Operating system:** Ubuntu 18.04

## Problem description
In using the `num-bigint` package (version 0.2.6, [source](https://docs.rs/crate/num-bigint/0.2.6)), the Rust plugin incorrectly labels subtracting a `u16` from a `&BigUint` as a type mismatch, when in reality the appropriate conversion and operation traits exist to do so. I believe this is because they are implemented using macros rather than directly, and this causes the Rust linter to not correctly detect the implementations.

## Steps to reproduce
Given `Cargo.toml`:
```toml
[package]
name = "bigint-sample"
version = "0.1.0"
authors = ["Eleanor McMurtry <foo@bar.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
num-bigint = "0.2.6"
```
The below code reproduces the problem:
```rust
use num_bigint::BigUint;

fn main() {
    let p = BigUint::from(13u64);
    let k = p.clone() - 1u16; // ok
    let k1 = &p - 1u16; // underlined "mismatched types [E0308]: expected `BigUint`, found `u16`
    let k2 = &p - &1u16; // underlined "mismatched types [E0308]: expected `BigUint`, found `&u16`

    println!("{}", k);  // outputs 12
    println!("{}", k1); // outputs 12
    println!("{}", k2); // outputs 12
}
```

A relevant excerpt from the `num-bigint` source code is included below -- there is a chain of macros used to lift conversions from `u32`. Note I'm not very familiar with the implementation so I have not been able to fully trace through and analyse what might be causing it.
```rust
promote_unsigned_scalars!(impl Sub for BigUint, sub);
promote_unsigned_scalars_assign!(impl SubAssign for BigUint, sub_assign);
forward_all_scalar_binop_to_val_val!(impl Sub<u32> for BigUint, sub);
forward_all_scalar_binop_to_val_val!(impl Sub<u64> for BigUint, sub);
#[cfg(has_i128)]
forward_all_scalar_binop_to_val_val!(impl Sub<u128> for BigUint, sub);

impl Sub<u32> for BigUint {
    type Output = BigUint;

    #[inline]
    fn sub(mut self, other: u32) -> BigUint {
        self -= other;
        self
    }
}
impl SubAssign<u32> for BigUint {
    fn sub_assign(&mut self, other: u32) {
        sub2(&mut self.data[..], &[other as BigDigit]);
        self.normalize();
    }
}
```

## Similar issues
* I found [https://github.com/intellij-rust/intellij-rust/issues/3882](https://github.com/intellij-rust/intellij-rust/issues/3882), which says that `cfg`   attributes were not supported, but has been apparently fixed.

**Comment by vlad20012** [comment]
Just enable the new macro expansion engine as explained in #3628

**Comment by noneuclideangirl** [comment]
Perfect — I couldn't seem to find it! I'll double check that it works tomorrow.

**Comment by noneuclideangirl** [comment]
Thanks, my apologies for the incorrect flag. :)
